### PR TITLE
Support for insert startupify list in order, support for solaire mode and more.

### PR DESCRIPTION
**Pull request by DevelopmentCool2449**
Hello,
I made this patch that include some
features to dashboard:

- Support for insert startupify list in order,
   this adds more customization to dashboard.
- Support for solaire-more.
- Bugs fix to some defcustom types.

I would like to hear your thought and reviews about this patch.
Thanks.

### CODE REVIEW

File: dashboard-widgets.el [hunk 1]
```diff
@@ -342,6 +342,11 @@ ARGS should be a plist containing `:height', `:v-adjust', or `:face' properties.
   :type 'string
   :group 'dashboard)
 
+(defcustom dashboard-heading-shorcut-key-str " (%s)"
```
Comment (ricardoricho) [inline dashboard-widgets.el:345]: Naming is hard, maybe a more descriptive name could be `dashboard-heading-shorcut-format`. This could be another PR.

File: dashboard.el [hunk 2]
```diff
@@ -118,6 +118,19 @@
...
 
+(defcustom dashboard-startupify-list
+  '((banner    . dashboard-insert-banner)
+    (new-line . (lambda () (insert "\n")))
+    (navigator . dashboard-insert-navigator)
+    (items     . dashboard-insert-items)
+    (new-line . (lambda () (insert "\n")))
+    (footer . dashboard-insert-footer)
+    (new-line . (lambda () (insert "\n")))
+    (init-info . dashboard-insert-init-info))
```
Comment (jcs090218) [inline dashboard.el:129]: The default value should be:

```elisp
      '((banner    . dashboard-insert-banner)
        (new-line  . (lambda () (insert "\n")))
        (navigator . dashboard-insert-navigator)
        (new-line  . (lambda () (insert "\n")))
        (init-info . dashboard-insert-init-info)
        (items     . dashboard-insert-items)
        (new-line  . (lambda () (insert "\n")))
        (footer    . dashboard-insert-footer))
```

File: dashboard.el [hunk 3]
```diff
@@ -418,44 +465,23 @@ Optional argument ARGS adviced function arguments."
     (when recentf-is-on
       (setq recentf-list (dashboard-subseq recentf-list dashboard-num-recents)))
     (dashboard--with-buffer
+     (if (fboundp 'solaire-mode)
+          (solaire-mode t))
```
Comment (jcs090218) [inline dashboard.el:469]: Use `when` instead of `if`.

**Comment by jcs090218** [comment]
I like the idea and make this even more customizable. 🚀 

You are missing one separator at the end, see comparison below:

| Before | After |
|---|---|
| before | after |

And, one byte-compile warning:

```
In dashboard-insert-startupify-lists:
dashboard.el:464:10: Warning: Unused lexical variable ‘max-line-length’
```

File: dashboard-widgets.el [hunk 4]
```diff
@@ -370,10 +375,13 @@ nil then no banner is displayed."
 Will be of the form `(list-type . list-function)'.
 Possible values for list-type are: `recents', `bookmarks', `projects',
 `agenda' ,`registers'."
-  :type  '(repeat (alist :key-type symbol :value-type function))
+  :type  '(alist :key-type symbol :value-type function)
   :group 'dashboard)
 
-(defcustom dashboard-projects-backend 'projectile
+(defcustom dashboard-projects-backend
+  (if (fboundp 'projectile-find-file)
```
Comment (ricardoricho) [inline dashboard-widgets.el:382]: I don't think it is a good idea to "guess" custom variables, maybe it's time to change the default value to `project-el` because now it's included in Emacs. This would be another PR.

File: dashboard.el [hunk 5]
```diff
@@ -118,6 +118,19 @@
   :type 'boolean
   :group 'dashboard)
 
+(defcustom dashboard-startupify-list
+  '((banner    . dashboard-insert-banner)
```
Comment (ricardoricho) [inline dashboard.el:122]: I think it's a good idea to have this list, makes the dashboard more flexible.
 
We don't need the "name" part of the list, could be just the function.
For the new-line part, maybe we could define a `dashboard-insert-new-line` with an optional number of new lines to insert, and another `dashboard-insert-page-separator` that inserts the "page-separator"

**Review by ricardoricho — CHANGES_REQUESTED** [review]
Hi @DevelopmentCool2449 thanks for your PR, well maybe we can split it into two or three PRs. Let's focus this on the good idea of the "startupify list".

**Comment by DevelopmentCool2449** [comment]
> I like the idea and make this even more customizable. 🚀
> 
> You are missing one separator at the end, see comparison below:
> Before 	After
> before 	after
> 
> And, one byte-compile warning:
> 
> ```
> In dashboard-insert-startupify-lists:
> dashboard.el:464:10: Warning: Unused lexical variable ‘max-line-length’
> ```

Thanks i've fixed them.

**Comment by DevelopmentCool2449** [comment]
> Hi @DevelopmentCool2449 thanks for your PR, well maybe we can split it into two or three PRs. Let's focus this on the good idea of the "startupify list".

> Naming is hard, maybe a more descriptive name could be dashboard-heading-shorcut-format. This could be another PR.

Thanks, i'm bad naming variables, it's fixed.
Also I dont think this should be another PR, this is added to allow users apply faces to the string.

> I don't think it is a good idea to "guess" custom variables, maybe it's time to change the default value to project-el because now it's included in Emacs. This would be another PR.

i don't think, to projectile users this can be useful.
Also i'm thinking this could be in other PR, what do you all think about?

> I think it's a good idea to have this list, makes the dashboard more flexible.
> 
> We don't need the "name" part of the list, could be just the function.

I was thinking the same.

> For the new-line part, maybe we could define a dashboard-insert-new-line with an optional number of new lines to insert, and another dashboard-insert-page-separator that inserts the "page-separator"

Isn't already supplied by dashboard-insert-page-break?.

File: dashboard.el [hunk 6]
```diff
@@ -407,55 +424,65 @@ Optional argument ARGS adviced function arguments."
...
-        (dashboard-num-recents (or (cdr (assoc 'recents dashboard-items)) 0))
-        (max-line-length 0))
+        (dashboard-num-recents (or (cdr (assoc 'recents dashboard-items)) 0)))
     (when recentf-is-on
       (setq recentf-list (dashboard-subseq recentf-list dashboard-num-recents)))
     (dashboard--with-buffer
+     (if (fboundp 'solaire-mode)
+         (if dashboard-use-solaire
+             (solaire-mode t)
+           (solaire-mode -1)))
```
Comment (jcs090218) [inline dashboard.el:472]: `solaire-mode` shouldn't be in this package. Users can achieve this by adding a hook to `dashboard-after-initialize-hook`.

**Comment by DevelopmentCool2449** [comment]
> n is unused here. You can just use (or n 1).

Thanks, i've fixed it.

> solaire-mode shouldn't be in this package. Users can achieve this by adding a hook to dashboard-after-initialize-hook.

I didn't know it, since in the README file doesn't indicate how to due with png backgrounds with solaire
(see: https://github.com/emacs-dashboard/emacs-dashboard/issues/179),
Maybe the README file must include how to add solaire to dashboard.

**Comment by jcs090218** [comment]
There is an error `dashboard.el:0:0: error: scan-error: (Unbalanced parentheses 16291 19182)` (https://github.com/emacs-dashboard/emacs-dashboard/actions/runs/8271644450/job/22632203133?pr=506#step:5:61).

**Comment by DevelopmentCool2449** [comment]
> There is an error `dashboard.el:0:0: error: scan-error: (Unbalanced parentheses 16291 19182)` (https://github.com/emacs-dashboard/emacs-dashboard/actions/runs/8271644450/job/22632203133?pr=506#step:5:61).

Weird, i dont know why it throws error, i'm checking it.

**Comment by DevelopmentCool2449** [comment]
> There is an error `dashboard.el:0:0: error: scan-error: (Unbalanced parentheses 16291 19182)` (https://github.com/emacs-dashboard/emacs-dashboard/actions/runs/8271644450/job/22632203133?pr=506#step:5:61).

Fixed

**Comment by DevelopmentCool2449** [comment]
I've added `dashboard-banner-logo-title` as a separate function from `dashboard-insert-banner`.

Some users can find it useful for their own setups:
examples:
https://github.com/konrad1977/welcome-dashboard/blob/main/screenshots/screenshot_1.png

https://github.com/konrad1977/welcome-dashboard
https://www.reddit.com/media?url=https%3A%2F%2Fpreview.redd.it%2Fgruvbox-theme-emacs-logo-v0-3fthokihn4oc1.png%3Fwidth%3D1080%26crop%3Dsmart%26auto%3Dwebp%26s%3D6602267ea2d7b2876e310174f51000899b39ed54

File: dashboard-widgets.el (hunk 4 above)
Comment (jcs090218) [reply dashboard-widgets.el:386]: I agree with @ricardoricho. I'm using the built-in `project.el` as well.

**Comment by jcs090218** [comment]
Can you document (to `README.md`) newly added variables?

- `dashboard-heading-shorcut-format`
- `dashboard-startupify-list`

**Comment by DevelopmentCool2449** [comment]
> Can you document (to `README.md`) newly added variables?
> 
>     * `dashboard-heading-shorcut-format`
> 
>     * `dashboard-startupify-list`

Sure, can i also add more details to README file? somethings like solaire-mode and better explanation to variables.

**Comment by DevelopmentCool2449** [comment]
I accidentally added a newline to at the end of dashboard-insert-init-info, i don't know if it can be annoying.
Apologies 😓

**Review by jcs090218 — APPROVED** [review]
LGTM. Thank you for your hard work! :)
